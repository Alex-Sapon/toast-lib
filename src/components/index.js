export ToastContainer from 'components/toastContainer';
