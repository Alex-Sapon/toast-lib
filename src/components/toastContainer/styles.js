import styled from 'styled-components';

export const Container = styled.div`
  width: 370px;
  position: fixed;
  top: ${props => props.theme.top}%;
  left: ${props => props.theme.left}%;
  right: ${props => props.theme.right}%;
  bottom: ${props => props.theme.bottom}%;
  transform: translate(${props => props.theme.translate}%);
  padding: 1rem;
  display: grid;
  row-gap: ${props => props.distance}px;
  grid-template-rows: max-content;
`;
