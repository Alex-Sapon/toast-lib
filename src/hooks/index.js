export { useToast } from 'hooks/useToast';
export { usePortal } from 'hooks/usePortal';
export { useDraggable } from 'hooks/useDragabble';
