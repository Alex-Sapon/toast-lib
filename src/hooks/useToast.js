import { useCallback, useState } from 'react';

import { createToast } from '@/helpers/createToast';
import { ToastFactory } from '@/utils/toastFactory';

export const useToast = (amountToasts = 3) => {
  const [instance] = useState(() => ToastFactory.getInstance());

  const removeToast = useCallback(id => instance.remove(id), [instance]);

  const addToast = (message, options, type) => {
    if (instance.toasts.length === amountToasts) return;
    const toast = createToast(message, options, type);
    instance.add({ ...toast, remove: removeToast });
    removeToast(toast);
  };

  const toaster = {
    success: (message, options) => addToast(message, options, 'success'),
    info: (message, options) => addToast(message, options, 'info'),
    warning: (message, options) => addToast(message, options, 'warning'),
    error: (message, options) => addToast(message, options, 'error'),
  };

  return { toaster };
};
