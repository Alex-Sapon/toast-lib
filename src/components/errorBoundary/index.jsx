import React from 'react';

import { ErrorText } from './styles';

export class ErrorBoundary extends React.Component {
  constructor(props) {
    super(props);
    this.state = { hasError: false };
  }

  static getDerivedStateFromError() {
    return { hasError: true };
  }

  render() {
    const { hasError } = this.state;
    const { children } = this.props;

    if (hasError) {
      return <ErrorText>Something wrong...</ErrorText>;
    }

    return children;
  }
}
