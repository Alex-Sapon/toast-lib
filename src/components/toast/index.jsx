import React, { memo, useCallback, useEffect, useRef, useState } from 'react';

import PropTypes from 'prop-types';

import close from 'assets/close.svg';
import {
  Container,
  Description,
  Icon,
  TextWrapper,
  Title,
  ProgressBar,
  CloseButton,
  CloseImg,
} from 'components/toast/styles';
import { getAnimation } from 'helpers/getAnimation';
import { useDraggable } from 'hooks';

export const Toast = memo(props => {
  const {
    id, title, message, autoClose, remove, position, animationName, ...restStyles
  } = props;

  const [finallyAnimation, setFinallyAnimation] = useState(false);
  const element = useRef();

  const handleAnimationEnd = () => setFinallyAnimation(true);

  const handleClose = useCallback(() => remove(id), [remove, id]);

  useDraggable(element, handleClose);

  useEffect(() => {
    if (autoClose) {
      const timerId = setTimeout(() => {
        handleAnimationEnd();
        remove(id);
      }, autoClose);
      return () => clearTimeout(timerId);
    }
  }, [remove, id, autoClose, finallyAnimation]);

  return (
    <Container
      ref={element}
      styles={restStyles}
      autoClose={autoClose}
      animation={getAnimation(position, animationName)}
      onAnimationEnd={handleAnimationEnd}
    >
      <Icon styles={restStyles} />
      <TextWrapper>
        <Title>{title}</Title>
        <Description>{message}</Description>
      </TextWrapper>
      <CloseButton onClick={handleClose}>
        <CloseImg src={close} />
      </CloseButton>
      <ProgressBar autoClose={autoClose} />
    </Container>
  );
});

Toast.propsType = {
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  autoClose: PropTypes.number,
  remove: PropTypes.func.isRequired,
  position: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  backgroundColor: PropTypes.string.isRequired,
  icon: PropTypes.string.isRequired
};
