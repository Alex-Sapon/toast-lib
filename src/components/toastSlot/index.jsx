import React, { memo } from 'react';

import { SlotContainer } from '@/components/toastSlot/styles';

import { Toast } from '@/components';
import { coordinates } from '@/styles/coordinates';

export const ToastSlot = memo(({ toasts, distance, animationName }) => {
  if (!toasts.length) return null;

  return (
    <SlotContainer
      coordinates={coordinates[toasts[0].position]}
      distance={distance}
    >
      {toasts.map(toast =>
        <Toast
          {...toast}
          key={toast.id}
          animationName={animationName}
        />
      )}
    </SlotContainer>
  );
});
