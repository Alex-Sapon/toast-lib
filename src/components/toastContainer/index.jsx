import React, { useState, useEffect } from 'react';

import PropTypes from 'prop-types';
import ReactDOM from 'react-dom';

import { ToastSlot } from '@/components/toastSlot';
import { ErrorBoundary } from '@/components/errorBoundary';
import { POSITIONS } from '@/constants/positions';
import { usePortal } from '@/hooks/usePortal';
import { GlobalStyles } from '@/styles/global';
import { ToastFactory } from '@/utils/toastFactory';

export const ToastContainer = ({ distance, animationName }) => {
  const [instance] = useState(() => ToastFactory.getInstance());
  const [toasts, setToasts] = useState(() => instance.toasts);
  const { loaded } = usePortal();

  const slots = POSITIONS.map(position =>
    toasts.filter(toast => toast.position === position),
  );

  useEffect(() => {
    instance.listener(newToasts => setToasts(newToasts));
  }, [instance]);

  return loaded
    ? ReactDOM.createPortal(
      <ErrorBoundary>
        {slots.map((slot, i) =>
          <ToastSlot
            key={POSITIONS[i]}
            toasts={slot}
            distance={distance}
            animationName={animationName}
          />,
        )}
        <GlobalStyles />
      </ErrorBoundary>,
      document.getElementById('toast-portal'))
    : null;
};

ToastContainer.defaultProps = {
  distance: 15,
};

ToastContainer.propsType = {
  distance: PropTypes.number,
  animationName: PropTypes.string,
};
