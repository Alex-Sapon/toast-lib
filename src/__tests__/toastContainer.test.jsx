import React from 'react';

import { render, renderHook } from '@testing-library/react';

import { Toast } from '@/components';
import { ToastContainer } from '@/components/toastContainer';
import { useToast } from '@/hooks/useToast';

it('render empty the ToastContainer component', () => {
  const { container } = render(<ToastContainer />);

  expect(container).toMatchSnapshot();
});

it('check useToast hook', () => {
  const { result } = renderHook(() => useToast());

  expect(result.current.toaster.success).toBeInstanceOf(Function);
});

it('renders the Toast component', () => {
  const { container } = render(
    <Toast
      id="1"
      title="Success"
      message="Test success message"
      autoClose={3000}
      remove={() => {}}
      position="bottom-left"
      color=""
      backgroundColor=""
      icon=""
    />,
  );

  expect(container).toMatchSnapshot();
});