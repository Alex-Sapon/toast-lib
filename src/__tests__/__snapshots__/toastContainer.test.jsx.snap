// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Toast component 1`] = `
<div>
  <div
    class="sc-gswNZR hFUYdi"
    style="transform: translateX(0px)false;"
  >
    <div
      class="sc-dkrFOg kCYMOK"
    />
    <div
      class="sc-hLBbgP"
    >
      <div
        class="sc-eDvSVe eIhXbY"
      >
        Success
      </div>
      <div
        class="sc-jSUZER ccLvKR"
      >
        Test success message
      </div>
    </div>
    <button
      class="sc-gKPRtg eaYjm"
    >
      <img
        class="sc-iBYQkv gVqbcP"
        src=""
      />
    </button>
    <div
      class="sc-ftTHYK eAhpLH"
    />
  </div>
</div>
`;

exports[`renders the ToastContainer component 1`] = `<div />`;
